### ISSUE TITLE
Use `cabal-install` instead of the machinery in `tests/test.hs`

### ISSUE THREAD

**Issue by Fizzixnerd** [ucsd-progsys/liquidhaskell#1955]
I have noticed that `tests/test.hs` is basically reproducing a build system, complete with manual dependency tracking and all the fun that entails.  I propose that we instead switch to using cabal.  I have a proof of concept on my for this with `tests/pos/`, which I will link to below.

## benefits of using cabal:

- automatic dependency tracking
- easier forced test suite running with forced rebuilds
- correct parallelism without hacks
- automatic tooling support! (editors will highlight liquid errors in test files, reducing the need to rerun the test suite after making changes and making development more interactive)
- more accurate and more detailed profiling via https://github.com/qnikst/ghc-timings-report (with `-ddump-timings` and `-ddump-to-file`)
- theoretically faster test running because of the lack of needing to startup ghc for each test file
- easier to add benchmarks (since they are just another cabal project)
- still supports partial test runs, but correctly connects dependencies (i.e., `cabal build MyFooTest` will build `MyFooLib` too (if needed!))
- much smaller footprint of `test.hs`, since it doesn't double as a build system; it only needs to parse the output of one

## complications:

- negative tests require a rework -- test.hs is still necessary to parse the output of cabal (or to use cabal as a library, whichever is easier!) for detecting specific error messages and the like
- need to adjust profiling infrastructure to new profile data -- see however ghc-timings-report/tmp/ in the below branch for the output of the new profiling tools.  Notice especially `tests-0.1.0.0.html` for a summary of how long the compiler took on each stage of compilation.
- Renames need to happen, but see `lh-renamer` in below branch for something that does _most_ of the work.
- legacy executable...

**Comment by Fizzixnerd** [comment]
https://github.com/Fizzixnerd/liquidhaskell/tree/test-hs-refactor2

**Comment by Fizzixnerd** [comment]
For simplicity's sake, here are screenshots of tests-0.1.0.0.htm
![Screenshot from 2022-03-27 13-40-50](https://user-images.githubusercontent.com/5892651/160293688-a296cf2d-de75-4d62-8fef-223c2fbca98a.png)
![Screenshot from 2022-03-27 13-40-54](https://user-images.githubusercontent.com/5892651/160293690-9092863d-5753-4cd9-a836-1412d8ec6f6b.png)
l:

**Comment by Fizzixnerd** [comment]
I did some preliminary profiling of the cabal build, and these are the results:

For the tests.cabal run,
```bash
time cabal build 2>/dev/null
```
gives
```
real	4m0.252s
user	4m8.265s
sys	0m38.609s
```
while for test.hs,
```bash
time stack --nix test --ta=-punit-pos
```
gives
```
real	8m59.245s
user	43m13.023s
sys	6m45.282s
```
on my machine.

**Comment by Fizzixnerd** [comment]
`time cabal build 2>/dev/null` with `-j` in `ghc-options` gives
```
real	2m12.028s
user	7m54.719s
sys	1m34.811s
```
as well

**Comment by Fizzixnerd** [comment]
`time cabal build 2>/dev/null` with `-j -O0` in `ghc-options` gives
```
real	1m0.848s
user	3m26.007s
sys	0m41.105s
```
which is pretty nice.

**Comment by Fizzixnerd** [comment]
I have extended the proof-of-concept to a benchmark (stitch-lh) and a couple other classes of tests.  I have also found the wonderful `-ddump-json` to turn GHC errors into json documents, which should make it much easier to parse the output of compilation for negative tests.  I need to ensure we know the requirements of the test-suite, so that we can be sure replacing it with cabal does not lead to loss of key functionality.

I believe next should be getting a negative suite to pass by modifying `test.hs`, and then integrating the legacy executable by co-opting cabal's preprocessor infrastructure to run liquid on each file.

**Comment by ranjitjhala** [comment]
It would be awesome to move to plain `cabal`! I am probably fine if we have to ditch the legacy executable to make this happen. Seems like its worth it. @nikivazou ?

**Comment by nikivazou** [comment]
Yes! 
The executable will still get generated, but not tested online correct?

**Comment by Fizzixnerd** [comment]
I have found a couple of issues that made it actually harder to parse the output with `-ddump-json` currently.  First, `aeson` only provides attoparsec parsers for json, and attoparsec parsers are not the most fun to debug,  Next, the plugin prints certain `WARNING:`s (when false is found in a specification, I believe?) and the headers `**** LIQUID: UNSAFE ************` etc as non-json, and also doesn't respect the `-fdiagnostics-color=never` flag, complicating the parser to be more than just `setBy endOfLine json'`.  I will be opening a couple of issues to track these, but they were not deal breakers.  Instead, I used megaparsec to write a parser for the raw output, which turned out to be pretty easy.

There was also issues with shelly: it seems to interleave stdout and stderr differently than it is when cabal is run directly.  I may have to switch away from it and use `typed-process` or the like instead.

I will upload my work soon, as my git index got clobbered somehow (worrisome, but probably my fault), but I wrote a summarizer that folds the list of compiler outputs into a count of the number of files which succeeded and failed.  It is quite simple to extend this to track the names of the files which succeeded/failed too, which I will then use to generate positive/negative reports.  The last remaining issue for the core functionality is to properly interleave stderr so that I can parse the liquid errors for the error-message-checking tests.  Currently, the parser is built to do this, but doesn't need to because it just operates on stdout which only receives the headers, so can only tell when Liquid is mad, and not why.

**Comment by Fizzixnerd** [comment]
Regarding abandoning the liquid executable testing: I would hold off on committing to this until it's clear it would be extra work to get it to work.  If I can successfully use the preprocessor infrastructure of cabal to run `liquid` as a preprocessor on each file before compiling normally with GHC, it may be possible to preserve the testing of it in the cabal framework.  However, I have doubts this will be as easy as that, since the preprocessor needs to run before GHC can do dependency analysis, yet the `liquid` executable requires this analysis to be done in order to work properly.  We shall see.

@nikivazou to answer your question, yes, if abandoned, the executable can be generated but not tested.  I am unsure if this is best in the long run, since shipping untested software is generally not a good idea, and if it's not being tested then it should probably be removed from the codebase and allow you to reap the benefits of such a simplification via refactoring.  But we can discuss as the issue arises, and I believe I am able to do whichever you choose to be best easily enough.

**Comment by Fizzixnerd** [comment]
See https://github.com/Fizzixnerd/liquidhaskell/blob/test-hs-refactor2/tests/harness/Test2.hs for the new test harness as I write it.

**Comment by Fizzixnerd** [comment]
I'm having an issue with the text benchmark: When I try to compile it with cabal, I get the following error:

```text
[15 of 39] Compiling Data.Text.Internal ( benchmarks/text-0.11.2.3/Data/Text/Internal.hs, /home/fizzixnerd/src/tweag/internal/liquidhaskell-redux/tests/dist-newstyle/build/x86_64-linux/ghc-8.10.7/tests-0.1.0.0/x/text-benchmark/build/text-benchmark/text-benchmark-tmp/Data/Text/Internal.o )

benchmarks/text-0.11.2.3/Data/Text/Internal.hs:138:15: error:
    • Bad Invariant Specification
    invariant  [{v0 : Data.Text.Internal.Text | len v > 1
                                                && tlen v0 > 0 => tlen v0 < sum_tlens v}]
    Sort Error in Refinement: {v0 : Data.Text.Internal.Text | len v > 1
                                                              && tlen v0 > 0 => tlen v0 < sum_tlens v}
    Unbound symbol v --- perhaps you meant: v0 ?
    • 
    |
138 | {-@ invariant {v:[{v0:Text | ((((len v) > 1) && ((tlen v0) > 0))
    |               ^

benchmarks/text-0.11.2.3/Data/Text/Internal.hs:140:15: error:
    • Bad Invariant Specification
    invariant  [{v0 : Data.Text.Internal.Text | len v > 0
                                                && tlen v0 > 0 => sum_tlens v > 0}]
    Sort Error in Refinement: {v0 : Data.Text.Internal.Text | len v > 0
                                                              && tlen v0 > 0 => sum_tlens v > 0}
    Unbound symbol v --- perhaps you meant: v0 ?
    • 
    |
140 | {-@ invariant {v:[{v0:Text | ((((tlen v0) > 0) && (((len v) > 0)))
```

where the offending lines are

```haskell
{-@ invariant {v:[{v0:Text | ((((len v) > 1) && ((tlen v0) > 0))
                                             => ((tlen v0) < (sum_tlens v)))}] | true} @-}
{-@ invariant {v:[{v0:Text | ((((tlen v0) > 0) && (((len v) > 0)))
                                             => ((sum_tlens v) > 0))}] | true} @-}
```

Just in case anyone has any ideas, I'm pretty stumped and am going to skip it for now and come back later.

**Comment by ranjitjhala** [comment]
I'm puzzled -- why does the old version work here
but not the new cabal-plugin that you're trying
[i.e. why are they behaving differently...]

(anyways, good idea to come back to it later!)

**Comment by Fizzixnerd** [comment]
> I'm puzzled -- why does the old version work here but not the new cabal-plugin that you're trying [i.e. why are they behaving differently...] (anyways, good idea to come back to it later!)

Reminder for future Matt: It could be related to the way I'm importing bytestring in the cabal file.

**Comment by Fizzixnerd** [comment]
(From yesterday, forgot to hit the green button!)

Today I made a couple of design decisions: first of all, I got rid of interleaving of stdout/stderr because it appeared to be non-deterministic and hard to make work properly.  Instead, I plan to split the output to two ByteStrings with the help of `typed-process`, and then use the stdout string primarily, doing lookups on stderr as needed.  This is easy enough to do because the error messages printed to stderr contain a filespan (i.e. something that looks like `pos/MyModule.hs:3:24`), and so we can do searches based on module name.

Second, I stripped all coloring from the output.  I'm not entirely sure why it took me so long to do this, but it has simplified the parsing and made it more deterministic as well, which is always a good thing.

I also updated the summarizer to only print the "failing" cases, instead of a giant list of everything that ran.  This has made the output much more pleasant to look at.

**Comment by facundominguez** [comment]
> I also updated the summarizer to only print the "failing" cases, instead of a giant list of everything that ran. This has made the output much more pleasant to look at.

Checking that tests are actually running is sometimes necessary, maybe the failing cases can be listed at the very end but still print everything. Otherwise, we will need to resort to flags or inspecting the generated reports.

**Comment by Fizzixnerd** [comment]
I finished the infrastructure for associating error messages with the modules that created them (remember, stderr is now a different stream!).  This means that when a file fails to build, the summarizer can now tell you why without you having to scroll through the output, which is nice.  More importantly, I can finally implement the `tests/error` type tests that are checked for specific error messages (since I now have those error messages!).

I also integrated with the CI on my fork.  See #1960 for the status.  I still need to catch a couple of weird conditions in the summarizer, but it mostly works (for cabal, at least).  I have decided to do stack by just stripping the `module-name > ` header from each line, but I haven't actually gotten around to it yet.

Here is an example of output from the summarizer, and see the attached picture for the colored output.  

![Screenshot from 2022-04-05 12-21-21 (cropped)](https://user-images.githubusercontent.com/5892651/161801017-cc63559f-8fd0-45e4-957c-77db2117127a.png)

```text

basic-neg:
  All good (8)
classes-neg:
  All good (4)
classes-pos:
  All good (2)
datacon-neg:
  All good (7)
measure-neg:
  All good (13)
names-neg:
  All good (9)
reflect-neg:
  All good (3)
unit-neg:
Unexpected
  42 / 194: Ex1_unsafe
    No Liquid Info
    
  96 / 194: MultiParamTypeClasses
    No Liquid Info
    
  119 / 194: Prune0
    No Liquid Info
    
  141 / 194: StrictPair1
    No Liquid Info
```

The Unexpected result for the `unit-neg` is from _passing_ tests, which I believe aren't using Liquid at all.  I haven't looked into it, in case it turns out they are ignored or something, but it's on my list.

**Comment by Fizzixnerd** [comment]
> > I also updated the summarizer to only print the "failing" cases, instead of a giant list of everything that ran. This has made the output much more pleasant to look at.
> 
> Checking that tests are actually running is sometimes necessary, maybe the failing cases can be listed at the very end but still print everything. Otherwise, we will need to resort to flags or inspecting the generated reports.

I will ensure everything is printed inline (I believe it is right now).  What I was mentioning was just the summary at the end.

**Comment by facundominguez** [comment]
Maybe make explicit in the output what the numbers refer too. e.g:
```
basic-neg:
  All good (8 modules checked)
```

**Comment by Fizzixnerd** [comment]
Here is an updated output of the summarizer. 

```text
absref-pos:
  All good (13 modules checked)
basic-neg:
  Unexpected (1/9 modules failed)
    7 / 9: Main
      Safe
      
    
basic-pos:
  All good (19 modules checked)
benchmark-stitch-lh:
  Unexpected (3/18 modules failed)
    14 / 18: Language.Stitch.LH.Monad
      No Liquid Info
      
    16 / 18: Language.Stitch.LH.Lex
      No Liquid Info
      
    18 / 18: Main
      No Liquid Info
      
    
classes-neg:
  Unexpected (1/5 modules failed)
    4 / 5: Main
      Safe
      
    
classes-pos:
  All good (9 modules checked)
datacon-neg:
  Unexpected (1/8 modules failed)
    6 / 8: Main
      Safe
      
    
datacon-pos:
  All good (15 modules checked)
errors:
  Unexpected (3/3 modules failed)
    1 / 83: AmbiguousInline
      Safe
      
    2 / 83: AmbiguousReflect
      Safe
      
    3 / 83: CyclicExprAlias0
      No Liquid Info
      
    
measure-neg:
  Unexpected (1/14 modules failed)
    11 / 15: Main
      Safe
      
    
measure-pos:
  All good (26 modules checked)
names-neg:
  Unexpected (1/10 modules failed)
    4 / 10: Main
      Safe
      
    
names-pos:
  All good (29 modules checked)
parser-pos:
  All good (14 modules checked)
reflect-neg:
  Unexpected (1/4 modules failed)
    2 / 4: Main
      Safe
      
    
reflect-pos:
  All good (4 modules checked)
unit-neg:
  Unexpected (1/194 modules failed)
    75 / 194: Main
      Safe
      
    
unit-pos:
  All good (533 modules checked)
```

I have added error message checking, but with a couple of caveats currently.  Many of the error checking examples seem to panic ghc.  This is a problem, since it kills the entire compilation, and prints an error message in yet another format I must parse (we're up to like 5 now).  However, it is necessary to be able to catch ghc panics, so I'm not exactly complaining, as it is a requirement to be able to report a panic.  I wonder if many of these messages should be panics though?  A panic is accompanied by a message claiming it is a bug in GHC, so I think normally they are reserved for bugs in implementations, not user facing error messages.  In a perfect world, a user shouldn't be able to cause a panic.  I may open an issue to track this separately.

Currently panics are not caught, so this is why `errors:` above says only 3 modules checked, when there should be like 83.  It panicked on CyclicExprAlias0.  This will be a tomorrow thing, while I think of the best way to handle restarting on the next file... Input is welcome!

The dummy Main functions need to be ignored.  I will also attack this tomorrow.  This is why the negative tests all fail with `Main -> Safe`.

**Comment by Fizzixnerd** [comment]
One more thing I wanted to bring up is the "complexity budget" I've been working with in my head.  I think it's important to just sort of lay it out there, since we don't want to make things worse with this change.

Line counting some source files, we find:
```text
[nix-shell:~/src/tweag/internal/liquidhaskell-redux]$ wc -l tests/harness/{*,Test/*}
     32 tests/harness/Driver_cabal.hs
     30 tests/harness/Driver_stack.hs
wc: tests/harness/Test: Is a directory
      0 tests/harness/Test
     75 tests/harness/Test/Build.hs
    142 tests/harness/Test/Groups.hs
    203 tests/harness/Test/Parse.hs
     60 tests/harness/Test/Summary.hs
    165 tests/harness/Test/Types.hs
    707 total
```
Also of note is the new cabal file:
```text
1490 tests/tests.cabal
```
while, for the old tests,
```text
[nix-shell:~/src/tweag/internal/liquidhaskell-redux]$ wc -l tests/Test.hs
1084 tests/test.hs
```

If we take line naive line count as a measure of complexity then, it would seem we have increased the complexity of the system greatly.  I would argue this is not the case, however.

First of all, a cabal file is miles (or kilometers) less complex than a Haskell source file.  Most of the cabal file is just listing the names of the modules that get compiled.  Similarly, `Test/Groups.hs` is just a list of the tests we want to run, and error messages we expect.  `Test/Types.hs` is a mix of `Pretty` instances and data declarations, and so is pretty low on the complexity scale too.  A program composed entirely of data declarations doesn't actually _do_ anything, after all.  The bulk of the complexity the lives in `Driver_*.hs`, and the remaining `Test/{Build,Parse,Summary}.hs`.  `Parse.hs` could be simplified in the future if #1957 can be resolved, as then it just becomes a JSON traversal.   `Build` and `Summary` are probably close to how complex they need to be, so I don't see them getting any shorter any time soon.  They are the "core" functionality of this mini-testframework.

My thesis has always been that a mini-build-system is an order of magnitude more complex to maintain than a mini-test-framework; this style of parsing is more or less easy.  Parallelism and dependencies are hard.

The question then becomes: is it robust?  The previous system was very sensitive to even filename changes, as this affected build order. This is difficult to judge, but I hope the way I wrote the parsers make it so failures and changes can be caught early.  This will be a key aspect of review, I think.

Comments and questions welcome as always.

**Comment by Fizzixnerd** [comment]
After #1968, I have been working on getting the error message tests to work properly.  Right now there is a complication with three or four of the error messages not having location information, and so messing up the parser by eating the next message that does in one big gulp.  This worries me slightly, but I have some ideas cooking away at a good way of getting around this.

Secondly, there is the only instance of Liquid outputting `**** LIQUID ERROR` as the header, and printing the error message to stdout instead of stderr.  **I consider this a bug**, but I'm curious what the original intent was for this.  @nikivazou and/or @ranjitjhala, **do you have any idea?**  If not, I should probably open an issue to track it.

Finally, even when printing error messages, often no header is printed for the file.  I am unsure if this is intentional, or a result of the particular way the panics were caught in #1968 .  This one was easy to fix, but it made me reconsider what the actual point of the headers are.  When I say header, I mean `**** LIQUID SAFE` or `UNSAFE` or `ERROR` or whatever.  In the case of an error or unsafety, an error is printed to stdout and compilation is aborted.  **Is this a vestigial fragment of how it was done pre-plugin?**  Given that this message 1. is always colored 2. doesn't respect the JSON settings, and 3. appears (to me!) to serve no real purpose, perhaps it would be **easier just to get rid of it** (obviously in a separate PR)?  Once again, I ask for your input @nikivazou and @ranjitjhala, as I do not claim to know the actual intent of these messages.  If those messages are removed (or fixed to respect compiler flags), there are a number of _theoretical_ benefits, including more uniform parsing with `-ddump-json` and easier enabling of `-j` for faster test running, but I make no promises yet, and I don't believe these reasons should be used to decide if the messages are still useful.

I'm going to put the error message tests on the back burner for now, and integrate the rest of the benchmarks, as this is higher priority at the moment, I believe.  It will also give me a chance to think of how to approach the error message failures in my subconscious.  Finally, I need to ensure everything works with `stack`, and then I will rewrite history to make review (hopefully) a little easier.

In an effort to keep the size of the PR down (as it is already getting to be quite large), I propose a subsequent PR for integrating the benchmarking graphing stuff from ghc-timings-report, and subsequent PRs for enabling `-j` for the tests (as this is currently complicated by the header printing) under the assumption the tests run no slower than currently (which we should know by tomorrow).  Input is welcome!

**Comment by Fizzixnerd** [comment]
I have disabled `--diff` in some of the benchmarks, because they would obviously interfere with the timings if enabled.  Let me know if this is an issue.

**Comment by Fizzixnerd** [comment]
There are a number of messages that are supposed to only be printed when `loggingVerbosity` is set to `Loud`.  It looks like certain flags raise the logging verbosity in addition to their named effect?  Is this documented somewhere, and can I turn it off?

**Comment by Fizzixnerd** [comment]
This is the state of the tests as of this morning; I had a little fun with making them go faster, but mostly have been adding tests that weren't included yet.  I have parallelized things and removed a bunch of loud warnings currently, just so I can fix up the core of the framework and hopefully prove it indeed can be faster.  Alas, the speed is not exactly what I had hoped for.   [cabal_810](https://app.circleci.com/pipelines/github/Fizzixnerd/liquidhaskell/110/workflows/d4ec3b09-0c6c-48b9-922c-b013fc6fb26a/jobs/398) builds in 6m9s and tests in 14m45s (not accounting for the failing tests that could take slightly longer).  This is compared to [develop](https://app.circleci.com/pipelines/github/Fizzixnerd/liquidhaskell/97/workflows/b280b958-b6f7-4ebf-83a5-5204c5ff371d/jobs/358) where building is 9m25s and testing is 14m3s. Tests can be parallelized further, by running them on separate circle runners, but that seems like overkill at this point, and is definitely outside the scope of *this* PR.

There are confounding variables at play here, in terms of VMs and caching.  For example, cabal_900 of that develop pipeline takes 30 minutes total, and yet ghc 9 is normally much faster at building.  (This suggests that the cache was reused when I reran the job that initially failed, which makes sense given what I know about how the cache works.)  More detailed measurements can wait for a subsequent PR and a small optimization pass.  I just wanted to make sure it's understood that it's reasonable to assume that speed is not necessarily going to *increase much* for this PR, but it shouldn't *decrease at all*.

The goal has always been to decrease the complexity of the system being used, however, and despite the enormous diff (which I will try to break up), I still believe the new system is 1) simpler than the old and 2) can be made simpler with a few small PRs in the future.  I will see what can be done about the enormousness of the diff; a colleague of mine suggested doing the renames of all the tests as a separate commit, so the "core" changes can be reviewed separately more easily, on a per-commit basis.  I will definitely be rewriting history.

Today I will be focusing on 1) getting the tests to pass on cabal, 2) getting the tests to pass on stack, and 3) prepping the PR for review.

```text
*** SUMMARY ***
absref-neg:
  All good (8 modules checked)
absref-pos:
  All good (13 modules checked)
basic-neg:
  All good (9 modules checked)
basic-pos:
  All good (19 modules checked)
benchmark-bytestring:
  All good (9 modules checked)
benchmark-cse230:
  All good (10 modules checked)
benchmark-esop2013:
  All good (8 modules checked)
benchmark-icfp15-neg:
  All good (4 modules checked)
benchmark-icfp15-pos:
  All good (18 modules checked)
benchmark-stitch-lh:
  All good (18 modules checked)
benchmark-text:
  Unexpected (1/33 modules failed)
    32 / 46: Data.Text
      benchmarks/text-0.11.2.3/Data/Text.hs:484:1:warning
          Rule "TEXT append -> fused" may never fire
            because ‘append’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘append’
          |
      484 | "TEXT append -> fused" [~1] forall t1 t2.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:527:1:warning
          Rule "TEXT last -> fused" may never fire
            because ‘last’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘last’
          |
      527 | "TEXT last -> fused" [~1] forall t.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:545:1:warning
          Rule "TEXT tail -> fused" may never fire
            because ‘tail’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘tail’
          |
      545 | "TEXT tail -> fused" [~1] forall t.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:565:1:warning
          Rule "TEXT init -> fused" may never fire
            because ‘init’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘init’
          |
      565 | "TEXT init -> fused" [~1] forall t.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:583:1:warning
          Rule "TEXT null -> fused" may never fire
            because ‘null’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘null’
          |
      583 | "TEXT null -> fused" [~1] forall t.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:618:1:warning
          Rule "TEXT compareN/length -> compareLength" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      618 | "TEXT compareN/length -> compareLength" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:623:1:warning
          Rule "TEXT ==N/length -> compareLength/==EQ" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      623 | "TEXT ==N/length -> compareLength/==EQ" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:628:1:warning
          Rule "TEXT /=N/length -> compareLength//=EQ" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      628 | "TEXT /=N/length -> compareLength//=EQ" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:633:1:warning
          Rule "TEXT <N/length -> compareLength/==LT" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      633 | "TEXT <N/length -> compareLength/==LT" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:638:1:warning
          Rule "TEXT <=N/length -> compareLength//=GT" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      638 | "TEXT <=N/length -> compareLength//=GT" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:643:1:warning
          Rule "TEXT >N/length -> compareLength/==GT" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      643 | "TEXT >N/length -> compareLength/==GT" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:648:1:warning
          Rule "TEXT >=N/length -> compareLength//=LT" may never fire
            because ‘length’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘length’
          |
      648 | "TEXT >=N/length -> compareLength//=LT" [~1] forall t n.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:789:1:warning
          Rule "TEXT justifyLeft -> fused" may never fire
            because ‘justifyLeft’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘justifyLeft’
          |
      789 | "TEXT justifyLeft -> fused" [~1] forall k c t.
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:942:1:warning
          Pattern match(es) are non-exhaustive
          In an equation for ‘concat_sumP’:
              Patterns not matched:
                  [] []
                  (_:_) []
          |
      942 | concat_sumP fun (t:ts) = concat_sumP_go fun 0 (t:ts)
          | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      benchmarks/text-0.11.2.3/Data/Text.hs:1122:1:warning
          Rule "TEXT replicate/singleton -> replicateChar" may never fire
            because ‘replicate’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘replicate’
           |
      1122 | "TEXT replicate/singleton -> replicateChar" [~1] forall n c.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1122:1:warning
          Rule "TEXT replicate/singleton -> replicateChar" may never fire
            because ‘singleton’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘singleton’
           |
      1122 | "TEXT replicate/singleton -> replicateChar" [~1] forall n c.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1181:1:warning
          Rule "TEXT take -> fused" may never fire
            because ‘take’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘take’
           |
      1181 | "TEXT take -> fused" [~1] forall n t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1212:1:warning
          Rule "TEXT drop -> fused" may never fire
            because ‘drop’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘drop’
           |
      1212 | "TEXT drop -> fused" [~1] forall n t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1237:1:warning
          Rule "TEXT takeWhile -> fused" may never fire
            because ‘takeWhile’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘takeWhile’
           |
      1237 | "TEXT takeWhile -> fused" [~1] forall p t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1271:1:warning
          Rule "TEXT dropWhile -> fused" may never fire
            because ‘dropWhile’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘dropWhile’
           |
      1271 | "TEXT dropWhile -> fused" [~1] forall p t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1308:1:warning
          Rule "TEXT dropWhileEnd -> fused" may never fire
            because ‘dropWhileEnd’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘dropWhileEnd’
           |
      1308 | "TEXT dropWhileEnd -> fused" [~1] forall p t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1512:1:warning
          Rule "TEXT splitOn/singleton -> split/==" may never fire
            because ‘splitOn’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘splitOn’
           |
      1512 | "TEXT splitOn/singleton -> split/==" [~1] forall c t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1512:1:warning
          Rule "TEXT splitOn/singleton -> split/==" may never fire
            because ‘singleton’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘singleton’
           |
      1512 | "TEXT splitOn/singleton -> split/==" [~1] forall c t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1703:1:warning
          Rule "TEXT count/singleton -> countChar" may never fire
            because ‘count’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘count’
           |
      1703 | "TEXT count/singleton -> countChar" [~1] forall c t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1703:1:warning
          Rule "TEXT count/singleton -> countChar" may never fire
            because ‘singleton’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘singleton’
           |
      1703 | "TEXT count/singleton -> countChar" [~1] forall c t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1816:1:warning
          Rule "TEXT isPrefixOf -> fused" may never fire
            because ‘isPrefixOf’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘isPrefixOf’
           |
      1816 | "TEXT isPrefixOf -> fused" [~1] forall s t.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1852:1:warning
          Rule "TEXT isInfixOf/singleton -> S.elem/S.stream" may never fire
            because ‘isInfixOf’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘isInfixOf’
           |
      1852 | "TEXT isInfixOf/singleton -> S.elem/S.stream" [~1] forall n h.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:1852:1:warning
          Rule "TEXT isInfixOf/singleton -> S.elem/S.stream" may never fire
            because ‘singleton’ might inline first
          Probable fix: add an INLINE[n] or NOINLINE[n] pragma for ‘singleton’
           |
      1852 | "TEXT isInfixOf/singleton -> S.elem/S.stream" [~1] forall n h.
           | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
      benchmarks/text-0.11.2.3/Data/Text.hs:897:48:error
          Liquid Type Mismatch
          .
          The inferred type
            VV : {v : Data.Text.Internal.Text | (tlength v == 0 <=> tlen v == 0)
                                                && tlength v == numchars (tarr v) (toff v) (tlen v)
                                                && numchars (tarr v) (toff v) 0 == 0
                                                && tlen v > 0
                                                && tlen v >= 0
                                                && tlength v >= 0
                                                && numchars (tarr v) (toff v) (tlen v) >= 0
                                                && numchars (tarr v) (toff v) (tlen v) <= tlen v}
          .
          is not a subtype of the required type
            VV : {VV : Data.Text.Internal.Text | tlen VV < Data.Text.Array.maLen arr}
          .
          in the context
            arr : {arr : (Data.Text.Array.MArray a) | GHC.Types.True
                                                      && Data.Text.Array.maLen arr == sum_tlens ts'
                                                      && sum_tlens ts' == Data.Text.Array.maLen arr
                                                      && Data.Text.Array.maLen arr > 0
                                                      && Data.Text.Array.maLen arr >= 0}
             
            ts : {ts : [Data.Text.Internal.Text] | len ts >= 0
                                                   && sum_tlens ts >= 0}
             
            ts' : {ts' : [Data.Text.Internal.Text] | sum_tlengths ts' == sum_tlengths ts
                                                     && sum_tlens ts' == sum_tlens ts
                                                     && len ts' >= 0
                                                     && sum_tlens ts' >= 0}
          Constraint id 506
          |
      897 |                                concat_step arr ts' 0 >> return arr
          |                                                ^^^
    
benchmark-vector-algorithms:
  All good (11 modules checked)
class-neg:
  All good (5 modules checked)
class-pos:
  All good (9 modules checked)
datacon-neg:
  All good (8 modules checked)
datacon-pos:
  All good (15 modules checked)
errors:
  Unexpected (11/101 modules failed)
    1 / 101: AmbiguousInline
      
    2 / 101: AmbiguousReflect
      
    16 / 101: BadQualifier
      
    39 / 101: ElabLocation
      
    44 / 101: ExportMeasure0
      
    45 / 101: ExportReflect0
      
    63 / 101: MissingField1
      
    78 / 101: ReflectLam
      
    79 / 101: ShadowFieldInline
      
    94 / 101: UnboundCheckVar
      
    95 / 101: UnboundFunInSpec
      
    The error messages of:
    39 / 101: ElabLocation
      
    didn't match expected error message of:
    ElabLocation.hs:13:14-13:15: Error
    The error messages of:
    79 / 101: ShadowFieldInline
      
    didn't match expected error message of:
    Multiple specifications for `pig`
    The error messages of:
    94 / 101: UnboundCheckVar
      
    didn't match expected error message of:
    Unknown variable `UnboundCheckVar.ink`
    The error messages of:
    95 / 101: UnboundFunInSpec
      
    didn't match expected error message of:
    Illegal type specification for `UnboundFunInSpec.three`
import-cli had fishy output from parser when attempting to scan stderr.
measure-neg:
  All good (14 modules checked)
measure-pos:
  All good (26 modules checked)
names-neg:
  All good (10 modules checked)
names-pos:
  All good (29 modules checked)
parser-pos:
  All good (14 modules checked)
ple-neg had fishy output from parser when attempting to scan stderr.
ple-pos had fishy output from parser when attempting to scan stderr.
prover-foundations:
  All good (5 modules checked)
prover-nople-neg:
  Unexpected (2/12 modules failed)
    9 / 12: Proves
      
    10 / 12: Helper
      
    
prover-nople-pos:
  All good (26 modules checked)
prover-ple-lib:
  All good (3 modules checked)
prover-ple-pos:
  All good (29 modules checked)
rankN-pos:
  All good (4 modules checked)
reflect-neg:
  All good (4 modules checked)
reflect-pos:
  All good (4 modules checked)
terminate-neg:
  All good (16 modules checked)
terminate-pos:
  All good (21 modules checked)
unit-neg:
  All good (194 modules checked)
unit-pos:
  All good (533 modules checked)
```

**Comment by Fizzixnerd** [comment]
GHC will occasionally segfault or exit with a double free error when invoked with `-j`.  I believe this is a bug in the Liquid plugin, but I haven't had time to investigate it properly.  This makes running tests in parallel a little perilous, unfortunately.  I have disabled `-j` on the tests that fail most often with this, and left it on for the others at the moment.

I have written documentation for all of the test-driver.  It is relatively extensive but please let me know if you want more detail.  I have rewritten history for the PR and it could potentially be started to be reviewed at this point, if @ranjitjhala or @nikivazou want to take a look. (If you can suggest other reviewers, that would also be helpful!)  See the PR #1960 for more details.  Basically, I did all the renaming (which is not always recognized as a rename by git's diff algorithm, unfortunately) in a single commit, and then added the rest of the testing framework on top of that.  When I get the chance I may investigate alternative diff algorithms and propose the reviewers use them, if I can get them to work, because right now GitHub chugs on my laptop when I attempt to review the big commit with the 17k lines "changed" (actually just copy-pasted with minor changes).

Cabal is almost working, I think I have two more tests to fix (possibly one only, at this point).  Stack is not working yet, but I don't anticipate much effort required there -- it should be as easy as stripping the output of the line header stack prints, and running the same parser.  :crossed_fingers: 

I will be taking tomorrow off for Easter.  Have a happy holidays, if you're doing the same!

**Comment by Fizzixnerd** [comment]
I was able to *almost* fix stack today, I'll see after lunch if I can grind out the rest of it; basically stack has this insane behavior where it just builds all the executables it can, even if you don't ask it to.  Obviously, in our case, it fails when doing the negative tests, and then exits in a huff.  I am going to think about ways around this, I know of at least one, but I hate it -- give each test it's own flag and only enable one at a time.  If I can't think of anything else, it will have to do.

Cabal 900 passes, indicating that we are nearing the end.  I believe an initial review can begin despite the CI not being fully green, should you be able to find the time, and would help me iron out any problems in a timely manner.  Also, the Appveyor is going to be failing because haskell.org has been down (or was, I haven't checked it recently) for a couple of days apparently, and so it can't access downloads.haskell.org.  I will continue working on getting the CI green and turn my attention to the review in the coming days.

**Comment by Fizzixnerd** [comment]
As #1960 has been merged, I am closing this issue.
